### PR TITLE
Restructuring/Review

### LINKED ISSUE #1: gnatgen new project name.derp is problematic
Running the following will be ok

``` bash
$ gnatgen new project name.with.dots.everywhere
```

But this will make the GPR file have a bunch of dots at the project name which probably won't be a good idea (not sure about what the specification says about this - it might be allowed for big projects, with subprojects that need building with other recipees). 

For now a quick fix would be to replace the dots with an underscore. Good beginner issue for whoever may want to tackle this.

### LINKED ISSUE #3: Incomplete help documentation
```
PS C:\Users\p-kel\Desktop\gnatgen\bin> .\gnatgen --help
Did not understand that
PS C:\Users\p-kel\Desktop\gnatgen\bin> .\gnatgen
Use:
gnatgen <action> [params]
  new project - to create a template of a new project
```

As seen, `--help` is not recognized and the help documentation that is printed out is incomplete, as the `README.md` is using "actions" such as `print` and `p`.

### PR DESCRIPTION
**Pull request by Entomy**
I've restructured the project a bit, since typically speaking parent packages should not depend on their children. The parsing model has been changed to something much more robust, specifically a stack-based try parser that calls the appropriate library function. Code has been cleaned up and conforms better to best practices, such as string comparisons through uppercase conversions.

There's considerably more error messages and more helpful error messages, as well as an (internal) elimination of exception handlers as these are extremely expensive. 

I've done my best to keep to the established style, although I've done this in multiple sittings so I probably broke that several times. Reformat as you see fit.

Closes #1 
Closes #3

### CODE REVIEW

**Comment by Entomy** [comment]
This also fixes Issues #1 and #3

File: .vscode/tasks.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
+{
```
Comment (psyomn) [inline .vscode/tasks.json:1]: I don't tend to like to add things like these, as I prefer to be tool agnostic. Is this necessary?

**Review by psyomn — COMMENTED** [review]
Thanks for the contributions! I'd like to merge this, but there are some conflicts, which I'd like to ask you about before doing anything (these are trivial): I think there's some conflicts on whitespacing. I can take care of these, but if you want, you could try rebasing as well! 

I'll be adding some more linting flags on the next iteration.

By the by, I believe you mentioned that you were interesting in packaging this. It might make sense to change the name of gnatgen to something else. How do you feel about this?

File: .vscode/tasks.json (hunk 1 above)
Comment (Entomy) [reply .vscode/tasks.json:1]: This isn't used to do anything special and just describes the build command to run. No it's not necessary though. I can add the directory to `.gitignore` if you want.

**Comment by Entomy** [comment]
I'm not surprised by whitespacing issues. I use tabs normally with Ada, and was trying to remember to use spaces instead. I can try rebasing, but as it's an unfamiliar approach I might mess it up.

I don't mind a name change. Might be a good idea since "gnatgen" seems to be a core GNAT component. Also sort of makes it seem like it's GNAT specific which it isn't and shouldn't be. Generating GNAT specific aspects, attributes, or pragmas seems like generating overly specific code.

**Comment by Entomy** [comment]
> I'll be adding some more linting flags on the next iteration.

On a somewhat related note, a file that I can add that is a standard, but not universally supported `.editorconfig` can enforce some style guidelines in editors that support it. Linting (gnatpp probably?) is certainly a great idea. `.editorconfig` can help enforce these ahead of time; at least force me to be using spaces for this project. As you said you didn't want specific tooling, I'll add it to the `.gitignore` if you don't want it included.

**Comment by Entomy** [comment]
> you could try rebasing as well!

Looks like GitHub just sees the new commits until the merge happens, as [e5bf83f22f7398b52faff72d192f8dfc83142977](https://github.com/psyomn/gnatgen/pull/6/commits/e5bf83f22f7398b52faff72d192f8dfc83142977) was added to the pull request. Would rebasing still be necessary?

**Comment by psyomn** [comment]
Could fix some of the issues, but no worries -- I can try and take care some of the conflicts. Let me know!

> On a somewhat related note, a file that I can add that is a standard, but not universally supported .editorconfig can enforce some style guidelines in editors that support it. Linting (gnatpp probably?) is certainly a great idea. .editorconfig can help enforce these ahead of time; at least force me to be using spaces for this project. As you said you didn't want specific tooling, I'll add it to the .gitignore if you don't want it included.

I mean adding some flags for linting through the `gpr` file. I'm unfarmiliar with `gnatpp`, but if you're confident it can do a lot of the boring stuff without getting in the way, go ahead and add it (maybe in a separate PR)!

**Comment by Entomy** [comment]
Well I did fix spacing. I believe the only remaining issues are from code restructuring, which you'd have to resolve.in

Well like I said, adding linting flags is still a good idea, especially as that's the only way to handle naming/casing conventions. As far as I know GPR linting just calls gnatpp under the hood. Ignore the `.editorconfig` stuff, linting will take care of style rules.

**Pull request closed without merging**